### PR TITLE
 *: make conversions between `string` and usernames more explicit 

### PR DESCRIPTION
**Pull request by knz**
Informs (but does not fix) #55396

Informs #54696
Informs  #55389

tldr: the conversions between "external" strings and internal
usernames was unprincipled, and it turns out, incorrect in some
cases. This patch cleans this up by introducing a strict conversion
API.

**Background**

CockroachDB currently performs case folding and unicode NFC
normalization upon receiving usernames, specifically usernames received
from as SQL login principals.

Internally, usernames are often—but not always—considered
pre-normalized for the purpose of comparisons, privilege checks, role
membership checks and the like.

Finally, sometimes usernames are reported "back to outside".
In error messages, log files etc, but also:

- in the SQL syntax produced by SHOW CREATE, SHOW SYNTAX etc.
- to generate OS-level file names for exported files.

**New API**

This patch introduces a new data type `security.SQLUsername`.
It is incomparable and non-convertible with the Go `string`.

Engineers must now declare intent about how to do the conversion:

- `security.MakeSQLUsernameFromUserInput` converts an "outside" string
  to a username.
- `security.MakeSQLUsernameFromPreNormalizedString` promises that
  its argument has already been previously normalized.
- `security.MakeSQLUsernameFromPreNormalizedStringChecked` also checks
  that its argument has already been previously normalized.

To output usernames, the following APIs are also available.

- `username.Normalized()` produces the username itself, without
  decorations. These corresponds to the raw string (after
  normalization).

- `username.SQLIdentifier()` produces the username in valid
  SQL syntax, so that it can be injected safely in a SQL
  statement.

- `(*tree.FmtCtx).FormatUsername()` takes a username and properly
  handles quoting and anonymization, like `FormatName()` does for
  `tree.Name` already.

Likewise, conversion from/to protobuf is now regulated, via the new
APIs `username.EncodeProto()` and `usernameproto.Decode()`.

**Problems being solved**

- the usernames "from outside" were normalized sometimes, *but not
  consistently*:

  1. they were in the arguments of CREATE/DROP/ALTER ROLE. This was
     not changed.
  2. they were not consistently converted in `cockroach cert`. This was
     corrected.
  3. they were not in the `cockroach userfile` commands. This
     has been adjusted with a reference to issue #55389.
  4. they are *not* in GRANT/REVOKE. This patch does not change
     this behavior, but highlights it by spelling out
     `MakeSQLUsernameFromPreNormalizedString()` in the implementation.
  5. ditto for CREATE SCHEMA ... AUTHORIZATION and ALTER ... OWNER TO
  6. they were in the argument to `LoginRequest`. This was not
     changed.
  7. they were not in the argument of the other API requests that
     allow usernames, for example `ListSessions` or `CancelQuery`.
     This was not changed, but is now documented in the API.

- the usernames "inside" were incorrectly directly injected
  in SQL statements, even though they may contain special
  characters that create SQL syntax errors.

  This has been corrected by suitable uses of the new
  `SQLIdentifier()` method.

- There was an outright bug in a call to `CreateGCJobRec` (something
  about GCing jobs), where a `Description` field was passed in lieu
  of a username for a `User` field. The implications of this are unclear.

**Status after this change**

The new API makes it possible to audit exactly where "sensitive"
username/string conversion occurs. After this patch, we find the
following uses:

- `MakeSQLUsernameFromUserInput`:

  - pgwire user auth
  - CLI URL parsing
  - `cockroach userfile`
  - `cockroach cert`
  - `(*rpc.SecurityContext).PGURL()` (unsure whether that's a good thing)
  - CREATE/DROP/ALTER ROLE
  - when using literal strings as `role_spec` in the SQL grammar

- `MakeSQLUsernameFromPreNormalizedString`:

  - role membership checks inside SQL based on data read
    from `system` tables.
  - in GRANT/REVOKE (this is surprising, see above)

- `MakeSQLUsernameFromPreNormalizedStringChecked`:

  - when intepreting the username in API query parameters, for
    those API documented as using pre-normalized usernames.

Release note: None

### CODE REVIEW

**Comment by knz** [comment]
NB: I'm not this change needs to be backported; however:

- @ajwerner @spaskob   I'd like you to check the fix to the `CreateGCJobRec` invocation and determine if the severity of this needs to be back-ported to 20.2.

- @miretskiy @dt I'd like you to check the generation of the `userfile_XXXX` special tables from usernames, and assess whether the existence of special usernames is severe enough of a bug we want to back-port the fix to that to 20.2.

Thanks

**Comment by dt** [comment]
IMO none of the userfile implications are severe enough to call for a 20.2.0 backport -- worse-case you get an error if the _default_ table name picked for your user is non-conforming, right? and then you just try again with a a valid table name explicitly specified? That seems _slightly annoying_ but IMO not a release blocker.

**Comment by knz** [comment]
Thanks David!


(Still interested in the implication for the schema change diff but maybe that's more a question for Andrew)

**Review by aaron-crl — APPROVED** [review]
This looks good modulo @solongordon 's feedback. I like pulling this principles into the `security` package as it's great hygiene and will help with tracking usage as @knz mentioned.

I have a nit around zero length usernames that I'd like others thoughts on. It shouldn't block this PR but I think it's worth discussing.

**Review by aaron-crl — COMMENTED** [review]
> I have a nit around zero length usernames that I'd like others thoughts on. It shouldn't block this PR but I think it's worth .

And it looks like reviewable ate my nit comment. I promise it was better written the first time.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55398#-:-MJSAr_G4GQOeeza5ygX:b-eulahp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @knz, @miretskiy, and @solongordon)

___
*[pkg/security/username.go, line 119 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55398#-MJSBhyU2bPfnWxgBpny:-MJSBhyU2bPfnWxgBpnz:b1ek1rv) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d6b20f7731c53f73e99f154843cf9528eeff88/pkg/security/username.go#L119)):*
> ```Go
> func MakeSQLUsernameFromUserInput(u string, purpose UsernamePurpose) (res SQLUsername, err error) {
> 	// Perform case folding and NFC normalization.
> 	res.u = lex.NormalizeName(u)
> ```

Nit: We still allow zero length usernames to be used within the application. Do we need/want to support this going forward?

It looks like NormalizeName will happily return a SQLUsername{""} if provided a zero length string and while we check length during user creation as part of the regex in ValidateForCreation() we don't check it outside of this case.

In addition Undefined() returns `true` when provided a zero length SQLUsername meaning that `user1.Undefined() == user2.Undefine()` can evaluate `true` which unsettles the math student in me.

Do we use/rely on this being able to accommodate a zero length field within the application? Given how many of the string tests we run will evaluate to vacuously true this makes me queasy. Would it be unreasonable to check min/max length in the constructor after normalization to protect against this case if we don't actively use this as a zero length value?

**Comment by knz** [comment]
I agree the validation function should check the username is non-empty. I will make this change.

**Review by arulajmani — COMMENTED** [review]
Change looks good, and thanks for laying the foundations for #54696 (among other issues). I left mostly minor comments and a couple questions (mostly for my understanding). I'll let @solongordon have the final review on this, though.

> Reviewed 158 of 231 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55398#-:-MJTJ_Bd7O3x2iKnMsVB:bw56jkj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @knz, @miretskiy, and @solongordon)

___
*[pkg/cli/cli_test.go, line 1392 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55398#-MJSXwJu8uoeDmSinauo:-MJSXwJu8uoeDmSinaup:biil05y) ([raw file](https://github.com/cockroachdb/cockroach/blob/d816ce972c4d28fcde15925ebc04e7c3522dbf20/pkg/cli/cli_test.go#L1392)):*
> ```Go
> 	// cert create-client ,foo
> 	// ERROR: failed to generate client certificate and key: username ",foo" invalid
> 	// SQLSTATE: 42602
> ```

Not sure if these comments have any bearing on the tests, but was this change intentional? 
___
*[pkg/security/username.go, line 128 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55398#-MJS_XHW7jNs1jwHYaVf:-MJS_XHW7jNs1jwHYaVg:bsl1fc3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d6b20f7731c53f73e99f154843cf9528eeff88/pkg/security/username.go#L128)):*
> ```Go
> // UsernamePurpose indicates the purpose of the resulting
> // SQLUsername in MakeSQLUsernameFromUserInput.
> type UsernamePurpose bool
> ```

nit: int (and iota below) instead of bool maybe?
___
*[pkg/security/username.go, line 180 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55398#-MJT7VMHAv_8Ie69T7fI:-MJT7VMHAv_8Ie69T7fJ:bs846rj) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d6b20f7731c53f73e99f154843cf9528eeff88/pkg/security/username.go#L180)):*
> ```Go
> // a canonical username. The caller of this promises that the
> // argument is pre-normalized.
> func MakeSQLUsernameFromPreNormalizedString(u string) SQLUsername {
> ```

Can we check the "pre-normalization" explicitly here instead of relying on the caller knowing what they're doing?
___
*[pkg/security/username.go, line 219 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55398#-MJSrEk-8TgaSUDxGZqF:-MJSrEk-8TgaSUDxGZqG:b-e8phi9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d6b20f7731c53f73e99f154843cf9528eeff88/pkg/security/username.go#L219)):*
> ```Go
> 
> // EncodeProto turns a username into its proto representation.
> func (u SQLUsername) EncodeProto() SQLUsernameProto { return SQLUsernameProto(u.u) }
> ```

super nit: my IDE is complaining about receiver names being different, can we change this to `s`? 
___
*[pkg/server/testserver.go, line 71 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55398#-MJScvBd-KADy3rNC0v1:-MJScvBd-KADy3rNC0v2:b1z7a3q) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d6b20f7731c53f73e99f154843cf9528eeff88/pkg/server/testserver.go#L71)):*
> ```Go
> 	// TestUser is a fixed user used in unittests.
> 	// It has valid embedded client certs.
> 	TestUser = security.TestUser
> ```

Can we get rid of this completely? I tried replacing all usages of `server.TestUser` with `security.TestUser` and didn't run into any dependency errors, so I wonder if we no longer need this. 
___
*[pkg/server/serverpb/status.proto, line 547 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55398#-MJTDfhrEUqfAR1bMh8r:-MJTDfhrEUqfAR1bMh8s:bemvu9j) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d6b20f7731c53f73e99f154843cf9528eeff88/pkg/server/serverpb/status.proto#L547)):*
> ```protobuf
>   // Username of the user making this request.
>   // The caller is responsible to normalize the username
>   // (= case fold and perform unicode NFC normalization).
> ```

Instead of relying on the caller, can we be a bit more prescriptive about this without breaking stuff? 

What if we replace the type of this to be `SQLUsernameProto` instead of `string`, and enforce that the username is normalized in the `EncodeProto` and `Decode` methods? 
___
*[pkg/sql/grant_revoke.go, line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55398#-MJT72Pb8VVpGKrcHRnP:-MJT72Pb8VVpGKrcHRnQ:bqqhatn) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d6b20f7731c53f73e99f154843cf9528eeff88/pkg/sql/grant_revoke.go#L65)):*
> ```Go
> 	// REASSIGN / OWNER TO do normalize.
> 	// Related: https://github.com/cockroachdb/cockroach/issues/54696
> 	grantees := make([]security.SQLUsername, len(n.Grantees))
> ```

What's the reason to consider not normalizing here? Backwards compat?

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55398)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aaron-crl, @arulajmani, @miretskiy, and @solongordon)

___
*[pkg/cli/cli_test.go, line 1392 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55398#-MJSXwJu8uoeDmSinauo:-MKA-skY-ZVTqpMIemQW:b-meyas9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d816ce972c4d28fcde15925ebc04e7c3522dbf20/pkg/cli/cli_test.go#L1392)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Not sure if these comments have any bearing on the tests, but was this change intentional? 
</blockquote></details>

yes, the comment _is_ the test (expected test output)

the `cert` commands are not SQL client commands, so having a SQLSTATE for them is nonsensical
___
*[pkg/security/username.go, line 119 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55398#-MJSBhyU2bPfnWxgBpny:-MKA0RZZDCowvtal1xz3:bask5v9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d6b20f7731c53f73e99f154843cf9528eeff88/pkg/security/username.go#L119)):*
> We still allow zero length usernames to be used within the application. 

No we don't. Added the check here.

> It looks like NormalizeName will happily return a SQLUsername{""} if provided a zero length string and while we check length during user creation as part of the regex in ValidateForCreation() we don't check it outside of this case.

Done

> In addition Undefined() returns true when provided a zero length SQLUsername meaning that user1.Undefined() == user2.Undefine() can evaluate true which unsettles the math student in me.

These comparisons do not occur in the source code.

> Do we use/rely on this being able to accommodate a zero length field within the application? 

No
___
*[pkg/security/username.go, line 128 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55398#-MJS_XHW7jNs1jwHYaVf:-MKA0hLW6zaHvHDmgOg-:b-6kwwxs) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d6b20f7731c53f73e99f154843cf9528eeff88/pkg/security/username.go#L128)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: int (and iota below) instead of bool maybe?
</blockquote></details>

Nah, as long as there are just 2 values bool just works fine.
___
*[pkg/security/username.go, line 180 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55398#-MJT7VMHAv_8Ie69T7fI:-MKA0pDYESmqoC-ew3XB:b-pn9mfd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d6b20f7731c53f73e99f154843cf9528eeff88/pkg/security/username.go#L180)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Can we check the "pre-normalization" explicitly here instead of relying on the caller knowing what they're doing?
</blockquote></details>

Discussed off the issue: I don't think we want this for performance reasons. The usernames fly over the wire for every SQL query.
___
*[pkg/security/username.go, line 219 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55398#-MJSrEk-8TgaSUDxGZqF:-MKA0vc_4xRwtltvo823:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d6b20f7731c53f73e99f154843cf9528eeff88/pkg/security/username.go#L219)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

super nit: my IDE is complaining about receiver names being different, can we change this to `s`? 
</blockquote></details>

Done.
___
*[pkg/server/testserver.go, line 71 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55398#-MJScvBd-KADy3rNC0v1:-MKA2s_S6H9LH8TL10A8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d6b20f7731c53f73e99f154843cf9528eeff88/pkg/server/testserver.go#L71)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Can we get rid of this completely? I tried replacing all usages of `server.TestUser` with `security.TestUser` and didn't run into any dependency errors, so I wonder if we no longer need this. 
</blockquote></details>

Done.
___
*[pkg/server/serverpb/status.proto, line 547 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55398#-MJTDfhrEUqfAR1bMh8r:-MKF-3Vs8Dy0wnDjOP0m:b-u5r28n) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d6b20f7731c53f73e99f154843cf9528eeff88/pkg/server/serverpb/status.proto#L547)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Instead of relying on the caller, can we be a bit more prescriptive about this without breaking stuff? 

What if we replace the type of this to be `SQLUsernameProto` instead of `string`, and enforce that the username is normalized in the `EncodeProto` and `Decode` methods? 
</blockquote></details>

Discussed this out of review: I am not keen to force checks in the proto conversion, for performance reason.

However, I think we can be prescriptive here by adding a check when receiving an API request. Done.
___
*[pkg/sql/grant_revoke.go, line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55398#-MJT72Pb8VVpGKrcHRnP:-MKF31aVAzbXzs8xCOM3:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d6b20f7731c53f73e99f154843cf9528eeff88/pkg/sql/grant_revoke.go#L65)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

What's the reason to consider not normalizing here? Backwards compat? 
</blockquote></details>

Fixed.

**Review by knz — COMMENTED** [review]
RFAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55398#-:-MKF57gC1Dq_AsYOXD3f:bb72pd0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aaron-crl, @arulajmani, @miretskiy, and @solongordon)

**Comment by knz** [comment]
Had to split package `lex` into `lexbase` / `lex` during rebase to avoid a cycle dependency due to the sessiondata refactor.

**Review by arulajmani — APPROVED** [review]
:lgtm: 

Could you also update the commit message to include a reference to `MakeSQLUsernameFromPreNormalizedStringChecked` as well? 

> Reviewed 6 of 69 files at r2, 53 of 88 files at r3, 11 of 15 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55398#-:-MKXGn5CBySokabjvi3i:b8zvrxj)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aaron-crl, @arulajmani, @knz, @miretskiy, and @solongordon)

___
*[pkg/sql/grant_revoke.go, line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55398#-MJT72Pb8VVpGKrcHRnP:-MKXFwjn8KGetHs_nF3N:b-rn5rsj) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d6b20f7731c53f73e99f154843cf9528eeff88/pkg/sql/grant_revoke.go#L65)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Fixed.
</blockquote></details>

Which layer was changed to fix this, I wasn't fully able to follow what happened here. 

Also, could you update the TODO (above and below)
___
*[pkg/sql/grant_revoke.go, line 163 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55398#-MKX011r4epHKCgZ2_U_:-MKX011r4epHKCgZ2_Ua:b-xse2a4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca1a0c89c55dc5d86319bcdecba2f681d1832e7e/pkg/sql/grant_revoke.go#L163)):*
> ```Go
> 		if _, ok := users[grantee]; !ok {
> 			sqlName := tree.Name(n.grantees[i].Normalized())
> 			return errors.Errorf("user or role %s does not exist", &sqlName)
> ```

nit: you can just use `grantee.Normalized()` here directly instead of this sqlName stuff

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55398)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aaron-crl, @arulajmani, @miretskiy, and @solongordon)

___
*[pkg/sql/grant_revoke.go, line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55398#-MJT72Pb8VVpGKrcHRnP:-MKaliqOE6f_uSM6zp6Q:b-mxs4bu) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d6b20f7731c53f73e99f154843cf9528eeff88/pkg/sql/grant_revoke.go#L65)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Which layer was changed to fix this, I wasn't fully able to follow what happened here. 

Also, could you update the TODO (above and below)
</blockquote></details>

Sorry I did not change anything here. So the TODO stands.
All in all this PR pushes the ball down the road on this question, which is why the linked issue remains open for now.
___
*[pkg/sql/grant_revoke.go, line 163 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55398#-MKX011r4epHKCgZ2_U_:-MKals9_CTzQ6LUBat9R:babkoli) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca1a0c89c55dc5d86319bcdecba2f681d1832e7e/pkg/sql/grant_revoke.go#L163)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: you can just use `grantee.Normalized()` here directly instead of this sqlName stuff
</blockquote></details>

No that's incorrect: tree.Name has different quoting rules, that of identifiers. for example the username `foo-bar` would quote as `"foo-bar"`. I did not want to change this behavior in this PR.

**Comment by knz** [comment]
> Could you also update the commit message to include a reference to MakeSQLUsernameFromPreNormalizedStringChecked as well?

done

**Comment by knz** [comment]
bors r=aaron-crl,arulajmani
